### ISSUE TITLE
configure: Better detection of installed nauty headers

### ISSUE THREAD

**Issue by mkoeppe** [Normaliz/Normaliz#332]
Normaliz' configure script seems to expect a `nauty/nauty.h` header file under the configured nauty prefix. However, the installation location of nauty header files seems to vary between distributions.

Conda-forge and Fedora-31 (libnauty-devel-2.6.12-1.fc31) have the file where Normaliz expects it and also has other header files in the same nauty/ subdirectory.
 
Homebrew (nauty-26r12) has  /usr/local/include/nauty.h (and no other headers).

Ubuntu bionic/focal and Debian unstable (libnauty2-dev-2.6r10+ds-1) have /usr/include/nauty/gtools.h and other header files in that directory, but no nauty.h

**Comment by w-bruns** [comment]
Normaliz uses densenauty, and this is defined in nauty.h. The nauty manual, Section 11, explicitly recommends its inclusion. Therefore a nauty installation without nauty.h is incomplete. There is nothing Normaliz can do if nauty.h is not present.

nauty has no make install. Therefore it is difficult to find a standard place for the header file.

Some packages install their headers in a subdirectory of include, others do not.  If you have an elegant solution how to overcome this problem, please let me know. It would certainly be possible to use a preprocessor variable like NO_NAUTY_INCLUDE_SUBDIR. I am sure one could insert a test in Normaliz' configure.ac for the location of nauty.h.

A pragmatic way out is to install nauty locally in the Normaliz directory tree. Normaliz only needs nauty.h and nauty.a under the name libnauty.a.

**Comment by bdmckay** [comment]
Even if nauty's makefile had an "install" target, it would still be the case that different systems put the files in different places.  To mkoeppe's list, add /opt/local/include for Macports.  So you either have to write configure.ac to search for nauty.h in lots of possible places, or you have to put it somewhere yourself during Normaliz build.  "./configure && make nauty.a" takes about 10 seconds, so the time cost is not an issue.  Making nauty.h and nauty.a as part of the Normaliz build would also mean that everyone gets the same version of nauty, in case that matters in the future.

Incidentally you probably should use "./configure --enable-wordsize=NN" where NN is the value you want (32 or 64), since the default value is not machine-independent.  That will define WORDSIZE to that value in nauty.h and build nauty.a with that value.

Cheers, Brendan.

**Comment by fingolfin** [comment]
Bundling nauty certainly would make some things easier for us. But I am also sure that e.g. Debian would rip it out, and modify our code to use their version of nauty after all. Which I am fine with (assuming it breaks nothing).

If we did that, we could remove the code from configure.ac for detecting nauty.

A variant would be to do what we did for GAP and GMP, and bundle a copy of nauty, but still offer the ability to link to another copy of nauty; then we can choose whether to always default to using the bundled nauty; or whether to default to first searching for an existing nauty, and then falling back to the bundled one.

Finally, we can also not bundle nauty and instead improve the build system to deal with various variants (which is not hard, but always annoying).

To summarize, I see these options, sorted in order of my personal preference (but I am willing to contribute towards implementing any of these); 1 and 2 are tied, but 1 is first because it requires less work ;-).

1. bundle nauty and remove build system support for linking to another nauty (of course packagers can work around that with moderate effort, should they want to)
2. bundle nauty, but keep supporting using another nauty version; with default to the bundled nauty
3. bundle nauty, but keep supporting using another nauty version; with default to an already installed nauty, if found
4. don't bundle nauty, but "enhance" the build system to deal with more cases

**Comment by mkoeppe** [comment]
I'll send a pull request to implement 4.

**Comment by mkoeppe** [comment]
Brendan,

> Even if nauty's makefile had an "install" target, it would still be the case that different systems put the files in different places. To mkoeppe's list, add /opt/local/include for Macports. 

The issue is specifically about whether `nauty.h` is installed in `PREFIX/include`, or in `PREFIX/include/nauty/nauty.h`.

That `PREFIX` is different on different installations is another matter, which Normaliz's `configure` script, of course, already handles.

I see that Debian has provided an autotoolization patch for nauty: https://sources.debian.org/patches/nauty/2.6r7+ds-1/
Any chance that upstream nauty will adopt this patch?

**Comment by w-bruns** [comment]
Max is certainly right about the amount of work necessary for the various options, but I would prefer 4. It is the most stable solution .

**Comment by mkoeppe** [comment]
Let's also submit bug reports to Debian regarding the missing nauty.h file
and to homebrew for possibly changing their install location to match that of the other major distributions.

**Comment by mkoeppe** [comment]
> I see that Debian has provided an autotoolization patch for nauty: https://sources.debian.org/patches/nauty/2.6r7+ds-1/

Current version: https://salsa.debian.org/science-team/nauty/-/tree/master/debian/patches

**Comment by mkoeppe** [comment]
> Let's also submit bug reports to Debian regarding the missing nauty.h file

Actually it's not missing. It is installed as `/usr/include/x86_64-linux-gnu/nauty/nauty.h` (this is due to Debian's multi-arch install scheme).
This change is done by https://salsa.debian.org/science-team/nauty/-/blob/master/debian/patches/debianization.patch

**Comment by mkoeppe** [comment]
> Let's also submit bug reports to [...] homebrew for possibly changing their install location to match that of the other major distributions.

The homebrew formula is here: https://github.com/Homebrew/homebrew-core/blob/master/Formula/nauty.rb

**Comment by w-bruns** [comment]
I lost control of what is wrong (or questionable) where.

**Comment by mkoeppe** [comment]
> I lost control of what is wrong (or questionable) where.
No action is needed at this time. I will provide a pull request when I have it ready.

**Comment by w-bruns** [comment]
Has the pull request been written? Version 3,.8.5 is about to be launched.

**Comment by mkoeppe** [comment]
OK, it does not look like we can expect a competently maintained package of nauty on homebrew.

**Comment by mkoeppe** [comment]
Pleasant surprise: Homebrew's nauty package has been updated in https://github.com/Homebrew/homebrew-core/pull/57159 and now has the header files in the same place as the other distributions. It only installs the static library though. But they do not plan to provide a shared library because upstream nauty does not build one.
```
==> Upgrading 1 outdated package:
nauty 26r12 -> 27r1
==> Upgrading nauty 26r12 -> 27r1 
...
(base) egret:~$ ls -l /usr/local/include/nauty/
total 392
-rw-r--r--  1 mkoeppe  staff      0 May 24 20:04 foo.h
-rw-r--r--  1 mkoeppe  staff  11450 May 24 20:04 gtools.h
-rw-r--r--  1 mkoeppe  staff   1910 May 24 20:04 gutils.h
-rw-r--r--  1 mkoeppe  staff   1553 May 24 20:04 naugroup.h
-rw-r--r--  1 mkoeppe  staff    527 May 24 20:04 naugstrings.h
-rw-r--r--  1 mkoeppe  staff   1157 May 24 20:04 naurng.h
-rw-r--r--  1 mkoeppe  staff   5721 May 24 20:04 nausparse.h
-rw-r--r--  1 mkoeppe  staff   1862 May 24 20:04 nautaux.h
-rw-r--r--  1 mkoeppe  staff   2470 May 24 20:04 nautinv.h
-rw-r--r--  1 mkoeppe  staff  15420 May 24 20:04 naututil.h
-rw-r--r--  1 mkoeppe  staff  58224 May 24 20:04 nauty.h
-rw-r--r--  1 mkoeppe  staff  17567 May 24 20:04 nautycliquer.h
-rw-r--r--  1 mkoeppe  staff  21373 May 24 20:04 planarity.h
-rw-r--r--  1 mkoeppe  staff  18148 May 24 20:04 quarticirred28.h
-rw-r--r--  1 mkoeppe  staff   1172 May 24 20:04 rng.h
-rw-r--r--  1 mkoeppe  staff   2741 May 24 20:04 schreier.h
-rw-r--r--  1 mkoeppe  staff   2977 May 24 20:04 traces.h
(base) egret:~$ ls -l /usr/local/lib/*nauty*
lrwxr-xr-x  1 mkoeppe  staff  35 Jul  2 11:32 /usr/local/lib/libnauty.a -> ../Cellar/nauty/27r1/lib/libnauty.a
```

**Comment by w-bruns** [comment]
The Normaliz install scripts only make and use the static library for the same reason.

**Comment by w-bruns** [comment]
The next release will be able to find "nauty/nauty.h" and naked "nauty.h". This feature exists already in the branch "simplified_reconciliation".
